proposals/p2240.md

```diff
@@ -656,90 +528,95 @@ starting with `Q` (corresponding to the cases where `x` and/or `z` are empty).
-expression cannot match a leading parameter, and a central argument expression
-can only match the variadic parameter.
-
-Consider the i'th non-variadic leading argument expression E. If all the
-variadic argument expressions before it are empty, E will match the i'th leading
-parameter, so E cannot match any earlier parameter. If there are any earlier
-variadic argument expressions, E can be made to match any later leading
-parameter or the variadic parameter, by making one of those earlier variadic
-arguments large enough, but as observed above, E cannot match a trailing
-parameter. If there are no earlier variadic argument expressions, E cannot be
-made to match any later parameter, so it can only match the i'th leading
-parameter.
-
-Next, consider a variadic leading argument expression E that comes before the
-i'th non-variadic leading argument expression, but not before any earlier
-non-variadic argument expression. If E's rewritten size is sufficiently large,
-and all earlier variadic argument expressions are empty, it will simultaneously
-match the i'th leading parameter, all leading parameters after it, and the
-variadic parameter, but as before, it cannot match a trailing parameter.
-
-The same reasoning can be applied to trailing argument expressions, but with
-"before" and "after" reversed. And as noted earlier, central argument
-expressions can only match the variadic parameter. In summary, we can identify
-the possible matches for an argument expression E as follows:
-
--   If E is leading, let i be one more than the number of earlier non-variadic
-    argument expressions:
-    -   If E is non-variadic, and there are no earlier variadic argument
-        expressions, then E can only match the i'th leading parameter.
+##### Identifying potential matchings
+
+Our solution will rely on being able to identify which pattern pack elements can
+potentially match which scrutinee pack elements. We can do so as follows:
+
+We will refer to the elements of the pack pattern as "parameters", and the
+elements of the scrutinee expression as "symbolic arguments". We will use the
+term "actual arguments" to refer to the elements of the scrutinee after
+monomorphization, so a single symbolic argument may correspond to any number of
+actual arguments, including zero (but only if the expression is variadic). Note
+that throughout this discussion, "parameters" and "arguments" are always types;
+we will sometimes say "parameter type" or "argument type" when there is a risk
+of confusion (FIXME there might still be too much risk of confusion, especially
+because we shift to more conventional meanings for "argument" and "parameter" in
+the next section). We will refer to arguments and parameters as "scalar" if they
+have size 1, and "variadic" if they have indeterminate size (these are the only
+possibilities, because packs are normalized).
+
+A pack pattern consists of N leading scalar parameters, optionally followed by a
```

Comment:
I don't think the below addresses the case where there is no variadic parameter.